File: su/src/main/java/org/sqlbuilder/su/Dump.java
package org.sqlbuilder.su;

import com.articulate.sigma.Formula;
import com.articulate.sigma.KB;

import java.io.PrintStream;
import java.util.Collection;
import java.util.List;

public class Dump
{
	public static void dumpTerms(final KB kb, final PrintStream ps)
	{
		int i = 0;
		for (final String term : kb.terms)
		{
			i++;
			ps.print("term " + i + "=" + term);
			ps.println(" doc=" + Dump.getDoc(kb, term));

			Dump.dumpParents(kb, term, ps);
			Dump.dumpChildren(kb, term, ps);
		}
	}

	public static void dumpParents(final KB kb, final String term, final PrintStream ps)
	{
		final Collection<Formula> formulas = kb.askWithRestriction(0, "subclass", 1, term);
		if (formulas != null && !formulas.isEmpty())
		{
			int i = 0;
			for (final Formula formula : formulas)
			{
				i++;
				final String formulaString = formula.getArgument(2);
				ps.print("\tparent" + i + "=" + formulaString);
				ps.println(" doc=" + Dump.getDoc(kb, formulaString));
			}
		}
	}

	public static void dumpChildren(final KB kb, final String term, final PrintStream ps)
	{
		final Collection<Formula> formulas = kb.askWithRestriction(0, "subclass", 2, term);
		if (formulas != null && !formulas.isEmpty())
		{
			int i = 0;
			for (final Formula formula : formulas)
			{
				i++;
				final String formulaString = formula.getArgument(1);
				ps.print("\tchild" + i + "=" + formulaString);
				ps.println(" doc=" + Dump.getDoc(kb, formulaString));
			}
		}
	}

	public static void dumpFormulas(final KB kb, final PrintStream ps)
	{
		int i = 0;
		for (final Formula formula : kb.formulas.values())
		{
			i++;
			ps.println(i + " " + formula);
		}
	}

	public static void dumpPredicates(final KB kb, final PrintStream ps)
	{
		final Collection<String> predicates = kb.collectPredicates();
		int i = 0;
		for (final String predicate : predicates)
		{
			i++;
			ps.println(i + " " + predicate);
		}
	}

	private static String getDoc(final KB kb, final String term)
	{
		final Collection<Formula> formulas = kb.askWithRestriction(0, "documentation", 1, term);
		if (formulas != null && !formulas.isEmpty())
		{
			final Formula formula = formulas.iterator().next();
			String doc = formula.getArgument(3);
			// doc = kb.formatDocumentation("http://", doc);
			doc = doc.replaceAll("\\n", "");
			return doc;
		}
		return null;
	}
}

File: su/src/main/java/org/sqlbuilder/su/Dump.kt
package org.sqlbuilder.su

import com.articulate.sigma.KB
import java.io.PrintStream

object Dump {

    private const val UP = "\uD83E\uDC45"

    private const val DOWN = "\uD83E\uDC47"

    fun dumpTerms(kb: KB, ps: PrintStream) {
        kb.terms.forEach {
            val doc = getDoc(kb, it)
            ps.println("$it $doc")
            dumpParents(kb, it, ps)
            dumpChildren(kb, it, ps)
        }
        ps.println(kb.terms.size)
    }

    fun dumpParents(kb: KB, term: String, ps: PrintStream) {
        kb.askWithRestriction(0, "subclass", 1, term)
            ?.withIndex()
            ?.forEach { (i, f) ->
                val formulaString = f.getArgument(2)
                ps.println("\t$UP $i $formulaString")
            }
    }

    fun dumpChildren(kb: KB, term: String, ps: PrintStream) {
        kb.askWithRestriction(0, "subclass", 2, term)
            ?.withIndex()
            ?.forEach { (i, f) ->
                val formulaString = f.getArgument(1)
                ps.println("\t$DOWN $i $formulaString")
            }
    }

    fun dumpFormulas(kb: KB, ps: PrintStream) {
        var count = 0
        kb.formulas.values
            .asSequence()
            .sorted()
            .forEach {
                ps.println(it)
                count++
            }
        ps.println(count)
    }

    fun dumpFunctions(kb: KB, ps: PrintStream) {
        var count = 0
        kb.collectFunctions()
            .asSequence()
            .sorted()
            .forEach {
                ps.println(it)
                count++
            }
        ps.println(count)
    }

    fun dumpRelations(kb: KB, ps: PrintStream) {
        var count = 0
        kb.collectRelations()
            .asSequence()
            .sorted()
            .forEach {
                ps.println(it)
                count++
            }
        ps.println(count)
    }

    fun dumpPredicates(kb: KB, ps: PrintStream) {
        var count = 0
        kb.collectPredicates()
            .asSequence()
            .sorted()
            .forEach {
                ps.println(it)
                count++
            }
        ps.println(count)
    }

    private fun getDoc(kb: KB, term: String): String {
        return kb.askWithRestriction(0, "documentation", 1, term)
            .asSequence()
            .sorted()
            .map {
                var doc = it.getArgument(3)
                doc.replace("\\n".toRegex(), "")

            }
            .joinToString(separator = "\n")
    }
}
